packages/client-node/src/result_set.ts

```diff
@@ -35,34 +38,64 @@ export class ResultSet implements BaseResultSet<Stream.Readable> {
 
     validateStreamFormat(this.format)
 
-    let decodedChunk = ''
+    let data: Buffer
+    let incompleteRow: Buffer | undefined
     const toRows = new Transform({
       transform(
         chunk: Buffer,
-        encoding: BufferEncoding,
+        _encoding: BufferEncoding,
         callback: TransformCallback
       ) {
-        decodedChunk += chunk.toString()
+        let currentDataSize
+        // The buffer was already allocated
+        // x2 chunk size should be enough for any leftover incomplete rows
+        if (data !== undefined) {
+          let chunkOffset
+          if (incompleteRow !== undefined) {
+            chunkOffset = incompleteRow.length
+            currentDataSize = chunkOffset + chunk.length
+            data.fill(incompleteRow, 0, chunkOffset)
+          } else {
+            chunkOffset = 0
+            currentDataSize = chunk.length
+          }
+          data.fill(chunk, chunkOffset, currentDataSize)
+          data.fill(0, currentDataSize, data.length)
+        } else {
+          // This is the first chunk, and we need to allocate enough memory
+          // to hold the chunk itself and any potential incomplete rows
+          currentDataSize = chunk.length
+          data = Buffer.alloc(chunk.length * 2, 0)
+          data.fill(chunk, 0, chunk.length)
+        }
         const rows: Row[] = []
-        // eslint-disable-next-line no-constant-condition
-        while (true) {
-          const idx = decodedChunk.indexOf('\n')
+        let lastIdx = 0
+        do {
+          // +1 to skip previously found '\n', if it's not the first pass
+          const startIdx = lastIdx === 0 ? 0 : lastIdx + 1
+          const idx = data.indexOf(NEWLINE, startIdx)
           if (idx !== -1) {
-            const text = decodedChunk.slice(0, idx)
-            decodedChunk = decodedChunk.slice(idx + 1)
+            const text = data.subarray(startIdx, idx).toString()
             rows.push({
               text,
               json<T>(): T {
-                return decode(text, 'JSON')
+                try {
+                  return JSON.parse(text)
+                } catch (e) {
+                  console.error('Failed with: ', text)
+                  throw e
+                }
               },
             })
           } else {
             if (rows.length) {
               this.push(rows)
             }
-            break
+            // Buffer after currentDataSize index is filled with zeroes
+            incompleteRow = data.subarray(startIdx, currentDataSize)
```

Comment:
Incomplete rows need to stack, as we can even have a row that does not fit in a single chunk. I will also be adding these checks.